src/Microsoft.ML.Data/DataLoadSave/Binary/Codecs.cs

```diff
@@ -1276,7 +1272,116 @@ private bool GetKeyCodec(Stream definitionStream, out IValueCodec codec)
                 Contracts.CheckDecode((ulong)count <= itemType.RawKind.ToMaxInt());
                 Contracts.CheckDecode(contiguous || count == 0);
 
-                type = new KeyType(itemType.RawKind, min, count, contiguous);
+                // Since we removed the notion of unknown cardinality (count == 0), we map to the maximum value.
+                if (count == 0)
+                    type = new KeyType(itemType.RawType, itemType.RawKind.ToMaxInt());
+                else
+                    type = new KeyType(itemType.RawType, count);
+            }
+            // Next create the key codec.
+            Type codecType = typeof(KeyCodecOld<>).MakeGenericType(itemType.RawType);
+            codec = (IValueCodec)Activator.CreateInstance(codecType, this, type, innerCodec);
+            return true;
+        }
+
+        private bool GetKeyCodecOld(ColumnType type, out IValueCodec codec)
```

Comment:
Does having `GetKeyCodecOld` mean we still use old key type attribute somewhere?

Reply:
This is necessary to allow backwards compatibility when reading files that were generated before this change. Take a look at the test that I added in BinaryLoaderSaverTests.cs in ML.Tests. Without maintaining the old codec, it would be impossible to read that file.

---
In reply to: [249909180](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249909180) [](ancestors = 249909180)

Reply:
Why do we need BC? Key type is changed to a new form.

---
In reply to: [249912390](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249912390) [](ancestors = 249912390,249909180)

Reply:
There is no reason to break backwards compatibility. We have I hope succeeded in never doing so. The entire reason for this change is that we never *used* these old capabilities. So if we have never used them, and we save data with this format (which is entirely possible), why would we break them? It is correct to maintain BC.

I do not view this as KeyType being changed to a new form at all. Rather, we are keeping only those parts of the KeyType have actually proved useful, and getting rid of the rest that just led to codebloat. I would not welcome any change that would break backcompat.